### ISSUE TITLE
Move/CopyAssignable should constrain the type of the assignment expression syntactically

### ISSUE THREAD

**Issue by CaseyCarter** [ericniebler/stl2#27]
In N4382, the MoveAssignable constrains the type of the assignment expression in prose:

> Let `rv` be an rvalue of type `T`. Then a type `T` models `MoveAssignable` if and only if
* The expression `t = rv` is an lvalue of type (non-`const`) `T`, the expression returns a reference to `t`, and after the assignment `t` is equal to the value of `rv` before the assignment.

the type requirement should be expressed in the syntax of the concept and removed from the prose:
```
template <class T>
concept bool MoveAssignable =
  requires(T a, T b) {
    { a = (T&&)b } -> Same<T&>;
  };
```

1 Let `rv` be an rvalue and `t` an lvalue of type `T`. The type `T` models `MoveAssignable` if and only if
* `(std::addressof(t = rv) == std::addressof(t)) != false`
* after the assignment `t = rv`, `t` is equal to the value of `rv` before the assignment.

**Comment by CaseyCarter** [comment]
CopyAssignable deserves the same treatment:
```
template <class T>
concept bool CopyAssignable =
  MoveAssignable<T> &&
  requires(T a, T b, T const& c) {
    { a = b } -> Same<T&>;
    { a = c } -> Same<T&>;
    { a = (T const&&)c } -> Same<T&>;
  };
```
1 Let `t` be an lvalue of type `T`, and `v` an lvalue of type (possibly `const`) `T` or an rvalue of type `const T`. The type `T` models `CopyAssignable` if and only if
* `(std::addressof(t = v) == std::addressof(t)) != false`,
* after the assignment `t` is equal to `v` and `v` is unchanged.

**Comment by ericniebler** [comment]
I'm curious why you wanted this change for Move/CopyAssignable, but not the other concepts like DefaultConstructible and friends.

**Comment by CaseyCarter** [comment]
Honestly, because I was looking at the Swappable requirements in N4431. The tables for MoveAssignable and CopyAssignable are on the same page, and looking over them in passing I thought the T& return type requirement was missing from N4382. When I checked, I realized I was wrong but decided it was odd that the requirement was in the prose instead of the concept syntax. I'll put the others under a magnifying glass as well and see if I can think of anything else that can move from semantics to syntax.
